### DISCUSSION TITLE
关于新增 InsertDict、UpdateDict、DeleteDict 功能的讨论

### DISCUSSION THREAD

**Discussion by hd2y** [dotnetcore/FreeSql#1055] [Ideas]
目前从示例来看，猜测方法声明为：

```cs
public IInsert<IDictionary<string, object>> InsertDict(this IFreeSql freeSql, IDictionary<string, object> dict);
public IInsert<IDictionary<string, object>> InsertDict(this IFreeSql freeSql, IEnumerable<IDictionary<string, object>> dictList);
// UpdateDict、DeleteDict 略
```

以下为一些个人建议。

+ 首先方法命名上倾向于使用 `XxxBy`、`XxxFrom`、`XxxUsing`：万一扩展字典以外类型可以复用这个名字，比如用户自行扩展一个方法，使用自己定义的 `CommonInsertModel<BaseEntity>` 封装当前 `InsertDict` 中需要配置的内容；
+ `AsTable` 增加泛型参数，在 `AsTable(string tbl)` 基础上增加重载，`AsTable<T>()`、`AsTable<T>(string tbl)`
  - 可以使用类型 `T` 定义 `Column` 特性的 `InsertValueSql` `IsPrimary` 等属性值
  - 如果解析到对应有导航属性被赋值，也执行新增操作等
+ 希望解析到字典 `object` 为 `JToken` 或者其他对象可以和 `JsonMap` 一样处理加入到数据库

cc @2881099

**Comment by 2881099** [comment]
1、为了与 IInsert\<T\>、IUpdate\<T\> 区分，目前是另外定义的新类型 InsertDictImpl、UpdateDictImpl 并且是以 IFreeSql 的扩展方法完成的该功能，原因是考虑到 IInsert/IUpdate 下的一些方法在 Dictionary\<string, object\> 类型下报错（无法完成），如 IgnoreColumn、Where、AsType，以及 BaseRepository\<Dictionary\<string, object\>\> 等等。

2、命名可以让更多人进行投票。

3、AsTable(string tbl) 已经增加了这个重载方法，并且 InsertDict/UpdateDict/DeleteDict 移掉了 AsTable(Func<string, string> tableRule) 方法，因为字典与同于实体类型，它本身不知道旧表名。

4、Column 特性需要进一步细节讨论。

5、导航属性，会不会让字典配置起来太复杂，需要进一步细节讨论。

**Reply by hd2y** [reply]
如果是这样，是否可以考虑提供一个 `InsertDictImpl<T>` 的抽象类，提供一个抽象方法 `IDictionary<string, object> GetDict(T data)`，默认 `InsertDictImpl: InsertDictImpl<IDictionary<string, object>>`。这样如果用户自己扩展，只需要实现该抽象方法，然后自己增加 `IFreeSql` 的扩展方法即可。

解析 `Column` 特性主要是看到示例中更新或者删除需要指定主键，如果 `AsTable` 增加带泛型参数的重载，这一步其实可以默认使用类型参数中定义的主键字段；另外就是有些数据插入时前端可能无法提供数据或者准确的数据，例如创建时间，之前考虑使用一个枚举来做映射，但是这倒不如 `InsertValueSql` 来得方便。

**Reply by 2881099** [reply]
1、IDictionary\<string, object\> GetDict(T data) 这个方法提供的作用是？

2、ColumnAttribute 专为实体提供设置，若同时用在字典会有诸多问题待解决。假设用户使用字典是 DbFirst 行为，那么它建表时就创建了默认值，插入数据时只要不设置字典对应的数据，是不是认为就完成了创建时间这种功能？
